util/proxy_protocol.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * util/proxy_protocol.c - event notification
+ *
+ * Copyright (c) 2022, NLnet Labs. All rights reserved.
+ *
+ * This software is open source.
+ * 
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 
+ * Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ * 
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * 
+ * Neither the name of the NLNET LABS nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ *
+ * This file contains PROXY protocol functions.
+ */
+#include "config.h"
+#include "util/log.h"
+#include "util/proxy_protocol.h"
+
+int
+pp2_write_to_buf(struct sldns_buffer* buf, struct sockaddr_storage* src,
+	int stream)
+{
+	int af;
+	if(!src) return 0;
+	af = (int)((struct sockaddr_in*)src)->sin_family;
+	if(sldns_buffer_remaining(buf) <
+		PP2_HEADER_SIZE + (af==AF_INET?12:36)) {
+		return 0;
+	}
+	/* sig */
+	sldns_buffer_write(buf, PP2_SIG, PP2_SIG_LEN);
+	/* version and command */
+	sldns_buffer_write_u8(buf, (PP2_VERSION << 4) | PP2_CMD_PROXY);
+	if(af==AF_INET) {
+		/* family and protocol */
+		sldns_buffer_write_u8(buf,
+			(PP2_AF_INET<<4) |
+			(stream?PP2_PROT_STREAM:PP2_PROT_DGRAM));
+		/* length */
+		sldns_buffer_write_u16(buf, 12);
+		/* src addr */
+		sldns_buffer_write_u32(buf,
+			htonl(((struct sockaddr_in*)src)->sin_addr.s_addr));
+		/* dst addr */
+		sldns_buffer_write_u32(buf, 0);
+		/* src port */
+		sldns_buffer_write_u16(buf,
+			htons(((struct sockaddr_in*)src)->sin_port));
```

Comment:
The port number in sockaddr_in.sin_port is already in network order and the sldns_buffer_write_u16 function applies htons on it.
```suggestion
			ntohs(((struct sockaddr_in*)src)->sin_port));
```

Reply:
I used `sldns_buffer_write(..., 2)` for this to avoid the conversions.